### ISSUE TITLE
Extremely slow for sorting asciidoc sections

### ISSUE THREAD

**Issue by mkschreder** [inkarkat/vim-AdvancedSorters#2]
**Describe the bug**

When I'm sorting folds using "'<,'>SortVisible /^== /" the operation takes forever. Undo takes even longer.

**How to Reproduce**

- Install habamax/vim-asciidoctor plugin.
- Enable folding: let g:asciidoctor_folding = 1
- Open an asciidoc and collapse all folds (zM)
- Visually select the sections you want to sort and then run the above command.

**Expected Behavior**

Should be quick.

**Environment**
 - Neovim 0.9.0

**Comment by inkarkat** [comment]
Yes, this can happen, not just with this plugin, but running substitutions on large texts that have complex folding or syntax definitions.

If there's any workaround (like `:set lazyredraw`, turning folds into manual ones, disabling syntax highlighting (all temporarily)) that reliably provides relief, I'd include such in the plugin.

**Comment by mkschreder** [comment]
@inkarkat I haven't looked exactly how the plugin works, but when it comes to sorting by a heading wouldn't it be possible to simply sort the list of references to all tags and then do just one substitution run at the end without having to do many insert/delete operations while sorting? Perhaps first sorting headings while saving their original position, and then moving the blocks around in some efficient manner so that they don't need to be moved more than once.

**Comment by inkarkat** [comment]
The plugin makes 3 separate changes: First it condenses each fold into a single line (joined via a NUL character), then invokes the built-in `:sort`, then expands the result back to the original ranges. That's the whole idea behind the plugin, and all of its commands are basically variations on that single theme.

That first join is a bit inefficient: It uses `:join!` and then needs to add the NUL character by temporarily going into insert mode. I've just pushed an update in inkarkat/vim-ingo-library@919a217 that instead uses `:substitute`. That might speed up the operation, but it could also not be noticeable at all, if a lengthy fold update or syntax parsing is still triggered for each line.
